src/core/net/mdns.hpp

```diff
@@ -0,0 +1,1118 @@
+        bool Contains(const Message &aMessage) const;
+        void HandleTimer(void);
+
+    private:
+        static constexpr uint32_t kExpireInterval = TimeMilli::SecToMsec(10); // in msec
+
+        typedef Crypto::Sha256::Hash Hash;
+
+        struct HashEntry : public LinkedListEntry<HashEntry>, public Heap::Allocatable<HashEntry>
+        {
+            bool Matches(const Hash &aHash) const { return aHash == mHash; }
+            bool Matches(const ExpireChecker &aExpireChecker) const { return mExpireTime <= aExpireChecker.mNow; }
+
+            HashEntry *mNext;
+            Hash       mHash;
+            TimeMilli  mExpireTime;
+        };
+
+        static void CalculateHash(const Message &aMessage, Hash &aHash);
+
+        using TxMsgHistoryTimer = TimerMilliIn<Core, &Core::HandleTxMessageHistoryTimer>;
+
+        OwningList<HashEntry> mHashEntries;
+        TxMsgHistoryTimer     mTimer;
+    };
+
+    // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+    template <typename EntryType> OwningList<EntryType> &GetList(void);
+    template <typename EntryType, typename ItemInfo>
+    Error Register(const ItemInfo &aItemInfo, RequestId aRequestId, RegisterCallback aCallback);
+    template <typename EntryType, typename ItemInfo> Error Unregister(const ItemInfo &aItemInfo);
+
+    void InvokeConflictCallback(const char *aName, const char *aServiceType);
+    void HandleMessage(Message &aMessage, bool aIsUnicast, const AddressInfo &aSenderAddress);
+    void RemoveEmptyEntries(void);
+    void HandleEntryTimer(void);
+    void HandleEntryTask(void);
+
+    static bool     IsKeyForService(const Key &aKey) { return aKey.mServiceType != nullptr; }
+    static uint32_t DetermineTtl(uint32_t aTtl, uint32_t aDefaultTtl);
+    static bool     NameMatch(const Heap::String &aHeapString, const char *aName);
+    static bool     NameMatch(const Heap::String &aFirst, const Heap::String &aSecond);
+    static void     UpdateCacheFlushFlagIn(ResourceRecord &aResourceRecord, Section aSection);
+    static void     UpdateRecordLengthInMessage(ResourceRecord &aRecord, Message &aMessage, uint16_t aOffset);
+    static void     UpdateCompressOffset(uint16_t &aOffset, uint16_t aNewOffse);
+    static bool     QuestionMatches(uint16_t aQuestionRrType, uint16_t aRrType);
+
+    using EntryTimer = TimerMilliIn<Core, &Core::HandleEntryTimer>;
+    using EntryTask  = TaskletIn<Core, &Core::HandleEntryTask>;
+
+    static const char kLocalDomain[];         // "local."
+    static const char kUdpServiceLabel[];     // "_udp"
+    static const char kTcpServiceLabel[];     // "_tcp"
+    static const char kSubServiceLabel[];     // "_sub"
+    static const char kServciesDnssdLabels[]; // "_services._dns-sd._udp"
+
+    bool                     mIsEnabled;
+    bool                     mIsQuestionUnicastAllowed;
+    uint16_t                 mMaxMessageSize;
+    uint32_t                 mInfraIfIndex;
```

Comment:
I'm guessing that mInfraIfIndex is an interface number. But it seems that is always used by the otPlatMdnsXyz() functions. If so why doesn't the platform implementation keeps track of the interface number?

Reply:
We want it to be similar to existing `otPlatInfraIf` APIs (where `aInfraIndex` is used in APIs). Including the infra-if index in the APIs will ensure we can extend/enhance the behavior in future.

Also please check related discussion in this comment https://github.com/openthread/openthread/pull/9797#discussion_r1466354688

Reply:
ok, got it; thanks